rust/basic_bitcoin/src/basic_bitcoin/src/bitcoin_wallet.rs

```diff
@@ -0,0 +1,259 @@
+//! A demo of a very bare-bones bitcoin "wallet".
+//!
+//! The wallet here showcases how bitcoin addresses can be be computed
+//! and how bitcoin transactions can be signed. It is missing several
+//! pieces that any production-grade wallet would have, including:
+//!
+//! * Support for address types that aren't P2PKH.
+//! * Caching spent UTXOs so that they are not reused in future transactions.
+//! * Option to set the fee.
+use crate::{bitcoin_api, ecdsa_api};
+use bitcoin::util::psbt::serialize::Serialize as _;
+use bitcoin::{
+    blockdata::script::Builder, hashes::Hash, Address, AddressType, OutPoint, Script, SigHashType,
+    Transaction, TxIn, TxOut, Txid,
+};
+use ic_btc_types::{Network, Satoshi, Utxo};
+use ic_cdk::{print, trap};
+use sha2::Digest;
+use std::str::FromStr;
+
+const SIG_HASH_TYPE: SigHashType = SigHashType::All;
+
+/// Returns the P2PKH address of this canister at the given derivation path.
+pub async fn get_p2pkh_address(network: Network, derivation_path: Vec<Vec<u8>>) -> String {
+    // Fetch the public key of the given derivation path.
+    let public_key = ecdsa_api::ecdsa_public_key(derivation_path).await;
+
+    // Compute the P2PKH address from our public key.
+    {
+        // sha256 + ripmd160
+        let mut hasher = ripemd::Ripemd160::new();
+        hasher.update(sha256(public_key));
+        let result = hasher.finalize();
+
+        let prefix = match network {
+            Network::Testnet | Network::Regtest => 0x6f,
+            Network::Mainnet => 0x00,
+        };
+        let mut data_with_prefix = vec![prefix];
+        data_with_prefix.extend(result);
+
+        let checksum = &sha256(sha256(data_with_prefix.clone()))[..4];
+
+        let mut full_address = data_with_prefix;
+        full_address.extend(checksum);
+
+        bs58::encode(full_address).into_string()
+    }
+}
+
+/// Sends a transaction to the network that transfers the given amount to the
+/// given destination, where the source of the funds is the canister itself
+/// at the given derivation path.
+pub async fn send(
+    network: Network,
+    derivation_path: Vec<Vec<u8>>,
+    dst_address: String,
```

Comment:
Could use [`Address`](https://gitlab.com/dfinity-lab/public/ic/-/blob/06f262a40a83a9a5aaa76b444e2225a6754448a1/rs/bitcoin/types/public/src/lib.rs#L7).

Reply:
I am using the types in the bitcoin crate only internally, but method signatures are using the (standard) address strings.